funnel/views/label.py

```diff
@@ -0,0 +1,194 @@
+            # If the label doesn't have any sublabel, one value is sent for each list,
+            # and those values are also available at `form.data`.
+            # But in case there are sublabels, the sublabel values are in the list
+            # in the order they appeared on the create form.
+            titlelist = request.values.getlist('title')
+            emojilist = request.values.getlist('icon_emoji')
+            # first values of both lists belong to the parent label
+            titlelist.pop(0)
+            emojilist.pop(0)
+
+            label = Label(title=form.data.get('title'), icon_emoji=form.data.get('icon_emoji'), project=self.obj)
+            label.restricted = form.data.get('restricted')
+            label.make_name()
+            self.obj.labels.append(label)
+            self.obj.labels.reorder()
+            db.session.add(label)
+
+            for idx, title in enumerate(titlelist):
+                subform = LabelOptionForm(MultiDict({
+                    'title': titlelist[idx], 'icon_emoji': emojilist[idx]
+                    }), meta={'csrf': False})  # parent form has valid CSRF token
+
+                if not subform.validate():
+                    flash(_("Error with a sublabel: {}").format(subform.errors.pop()), category='error')
+                    return dict(title="Add label", form=form, project=self.obj)
+                else:
+                    subl = Label(project=self.obj)
+                    subform.populate_obj(subl)
+                    subl.make_name()
+                    db.session.add(subl)
+                    label.options.append(subl)
+
+            db.session.commit()
+            return redirect(self.obj.url_for('labels'), code=303)
+        return dict(title="Add label", form=form, emptysubform=emptysubform, project=self.obj)
+
+
+@route('/<project>/labels', subdomain='<profile>')
+class FunnelProjectLabelView(ProjectLabelView):
+    pass
+
+
+ProjectLabelView.init_app(app)
+FunnelProjectLabelView.init_app(funnelapp)
+
+
+@route('/<profile>/<project>/labels/<label>')
+class LabelView(UrlForView, ModelView):
+    __decorators__ = [lastuser.requires_login, legacy_redirect]
+    model = Label
+    route_model_map = {
+        'profile': 'project.profile.name', 'project': 'project.name',
+        'label': 'name'}
+
+    def loader(self, profile, project, label):
+        proj = Project.query.join(Profile).filter(
+                Profile.name == profile, Project.name == project
+            ).first_or_404()
+        label = self.model.query.filter_by(project=proj, name=label).first_or_404()
+        g.profile = proj.profile
+        return label
+
+    @route('edit', methods=['GET', 'POST'])
+    @lastuser.requires_login
+    @render_with('labels_form.html.jinja2')
+    @requires_permission('edit_project')
+    def edit(self):
+        emptysubform = LabelOptionForm(MultiDict({}))
+        subforms = []
+        if not self.obj.main_label:
```

Comment:
This is confusing because for a moment I couldn't recall if `main_label` points to the main label, or is a property indicating that this is a main label (which should have been `is_main_label`). Let's add a property named `is_main_label` that simply returns `not self.main_label`.

Reply:
Or `is_top_level` to make it less ambiguous.